src/core/symbology/qgsfillsymbollayer.cpp

```diff
@@ -3509,41 +3513,118 @@ void QgsCentroidFillSymbolLayer::stopRender( QgsSymbolRenderContext &context )
 
 void QgsCentroidFillSymbolLayer::renderPolygon( const QPolygonF &points, QList<QPolygonF> *rings, QgsSymbolRenderContext &context )
 {
-  if ( !mPointOnAllParts )
+  Part part;
```

Comment:
There's a bug here (also in the random marker fill symbol layer) which causes issues in multi layer symbols. I can't merge this until that bug is fixed, because this will break any existing multi-layer symbols using the centroid fill (the severity of the issue with random marker fills isn't so bad, since the bug was present ever since the random marker symbol layer type was introduced... hence no regressions possible!)

Reply:
Should I close the PR then? Or just standby until the bug is fixed? I am afraid I am not qualified to track the bug (yet), so the only option for me is to wait or to change the implementation. Is there an open issue for the bug you mentioned?

Reply:
Standby for now -- what you have here looks good, it's deeper architectural changes required